crates/gitbutler-operating-modes/src/lib.rs

```diff
@@ -1,38 +1,113 @@
+use std::{fs, path::PathBuf};
+
 use anyhow::{bail, Context, Result};
 use gitbutler_command_context::CommandContext;
+use serde::{Deserialize, Serialize};
+
+pub const INTEGRATION_BRANCH_REF: &str = "refs/heads/gitbutler/integration";
+pub const EDIT_BRANCH_REF: &str = "refs/heads/gitbutler/edit";
 
-/// Operating Modes:
-/// Gitbutler currently has two main operating modes:
-/// - `in workspace mode`: When the app is on the gitbutler/integration branch.
-///     This is when normal operations can be performed.
-/// - `outside workspace mode`: When the user has left the gitbutler/integration
-///     branch to perform regular git commands.
+fn edit_mode_metadata_path(ctx: &CommandContext) -> PathBuf {
+    ctx.project().gb_dir().join("edit_mode_metadata.toml")
+}
 
-const INTEGRATION_BRANCH_REF: &str = "refs/heads/gitbutler/integration";
+pub fn read_edit_mode_metadata(ctx: &CommandContext) -> Result<EditModeMetadata> {
+    let edit_mode_metadata = fs::read_to_string(edit_mode_metadata_path(ctx).as_path())
+        .context("Failed to read edit mode metadata")?;
 
-pub fn in_open_workspace_mode(ctx: &CommandContext) -> Result<bool> {
+    toml::from_str(&edit_mode_metadata).context("Failed to parse edit mode metadata")
+}
+
+pub fn write_edit_mode_metadata(
+    ctx: &CommandContext,
+    edit_mode_metadata: &EditModeMetadata,
+) -> Result<()> {
+    let serialized_edit_mode_metadata =
+        toml::to_string(edit_mode_metadata).context("Failed to serialize edit mode metadata")?;
+    gitbutler_fs::write(
+        edit_mode_metadata_path(ctx).as_path(),
+        serialized_edit_mode_metadata,
+    )
+    .context("Failed to write edit mode metadata")?;
+
+    Ok(())
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct EditModeMetadata {
```

Comment:
Its not clear what these fields represent. Is the commit sha guaranteed to exist? What is it used for?  How about the branch? Is it just a label of some kind? Do you mean a reference? Why use string here?

Reply:
We have a tagged string reference type, maybe you wanna use that?

Reply:
> I can't use a reference because it's lifetime is tied to a git2::Repository (making it an ass pain) and can't be serialized/deserialized.

Right, this is good to know. `gitoxide` has fully owned versions, but I don't know if the serialize (yet). If not, that can easily be added, or own serializers could be added for it (then).

My stance these days is to use the strongest type possible in the backend, possibly converting to a specialized version just for the frontend. Using [`serde(flatten)` worked well for that](https://github.com/Byron/gitbutler/blob/39099ba41cd128b13642c72994096f44089feec7/crates/gitbutler-tauri/src/projects.rs#L97) previously.

Also, generally, I document everything I know about newly added types along with their fields, just because only in that moment I still know most about it.